cpp/src/arrow/util/async_generator.h

```diff
@@ -1063,6 +1063,86 @@ AsyncGenerator<T> MakeConcatenatedGenerator(AsyncGenerator<AsyncGenerator<T>> so
   return MergedGenerator<T>(std::move(source), 1);
 }
 
+template <typename T>
+struct Enumerated {
+  util::optional<T> value;
```

Comment:
I'm not sure why `util::optional<T>` is useful here. You can simply use `IterationTraits<T>::End()` when the end has been reached.

Reply:
Yes, that should work.  I'll change to that.